File: Languages/cpp/algorithms-dsa/leetcode/stack/valid_parentheses.cpp
/*
 * LeetCode 20 - Valid Parentheses
 * Topic: Stack
 * Difficulty: Easy
 *
 * Push opening brackets onto stack, pop and match for closing brackets.
 * Time: O(n)
 * Space: O(n)
 */
#include <iostream>
#include <string>
#include <stack>
#include <cassert>
using namespace std;

bool isValid(string s) {
    stack<char> st;
    for (char c : s) {
        if (c == '(' || c == '{' || c == '[') st.push(c);
        else {
            if (st.empty()) return false;
            char top = st.top(); st.pop();
            if ((c == ')' && top != '(') ||
                (c == '}' && top != '{') ||
                (c == ']' && top != '[')) return false;
        }
    }
    return st.empty();
}

int main() {
    assert(isValid("()") == true);
    assert(isValid("()[]{}") == true);
    assert(isValid("(]") == false);
    assert(isValid("([)]") == false);
    assert(isValid("{[]}") == true);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/stack/valid_parentheses.py
"""
LeetCode #20 - Valid Parentheses
Topic: Stack
Difficulty: Easy

Check if brackets are balanced using stack matching.

Time Complexity: O(n)
Space Complexity: O(n)
"""


def is_valid(s: str) -> bool:
    stack: list[str] = []
    pairs = {")": "(", "}": "{", "]": "["}
    for ch in s:
        if ch in pairs:
            if not stack or stack[-1] != pairs[ch]:
                return False
            stack.pop()
        else:
            stack.append(ch)
    return len(stack) == 0


if __name__ == "__main__":
    assert is_valid("()") is True
    assert is_valid("()[]{}") is True
    assert is_valid("(]") is False
    assert is_valid("([)]") is False
    assert is_valid("{[]}") is True
    print("All tests passed!")
